components/layout_thread/dom_wrapper.rs

```diff
@@ -157,35 +164,78 @@ impl<'ln> NodeInfo for ServoLayoutNode<'ln> {
     }
 }
 
-#[derive(Clone, Copy, PartialEq)]
-enum Impossible {}
+#[derive(Clone, Copy)]
+pub struct ServoShadowRoot<'a> {
+    /// The wrapped shadow root.
+    shadow_root: LayoutDom<ShadowRoot>,
+
+    /// Being chained to a PhantomData prevents `ShadowRoot`s from escaping.
+    chain: PhantomData<&'a ()>,
+}
 
-#[derive(Clone, Copy, PartialEq)]
-pub struct ShadowRoot<'lr>(Impossible, PhantomData<&'lr ()>);
+impl<'sr> Debug for ServoShadowRoot<'sr> {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        self.as_node().fmt(f)
+    }
+}
+
+impl<'a> PartialEq for ServoShadowRoot<'a> {
+    #[inline]
+    fn eq(&self, other: &ServoShadowRoot) -> bool {
+        self.shadow_root == other.shadow_root
+    }
+}
 
-impl<'lr> TShadowRoot for ShadowRoot<'lr> {
-    type ConcreteNode = ServoLayoutNode<'lr>;
+impl<'sr> TShadowRoot for ServoShadowRoot<'sr> {
```

Comment:
`l` stands for lifetime, so I'd prefer to keep `lr` for consistency with the other `le` and co.